File: CFromSwift/into.swift
import Foundation

typealias CString = UnsafePointer<CChar>

@_extern(c)
func sum(_ a: Int32, _ b: Int32) -> Int32

@_extern(c, "string_from_c")
func stringFromC() -> CString

@_extern(c, "c_strlen")
func cStrlen(_ str: CString) -> Int32

@_extern(c, "c_printstr")
func cPrintStr(_ str: CString)


@main
enum main {
    static func main() {
        print("Hello, World!\n\(sum(34, 35)) times"); 
        let cStrPtr = stringFromC()
        let cStr = String(cString: cStrPtr)
        let msg = "\"\(cStr)\"\nC says it's \(cStrlen(cStrPtr)) chars long\nwhile I say it's \(cStr.count) chars long"
        
        cPrintStr(msg.cStr())
        cPrintStr(stringFromC())
    }
}


extension String {
    func cStr() -> CString {
        CString(strdup(self))
    }
}

File: SwiftFromC/into.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int sum(int a, int b);
char* string_from_swift();
int swift_strlen(char *str);
void swift_printstr(char *str);


int main() {
    printf("Hello, World!\n%d times\n", sum(34, 35)); 
    char *swft_str = string_from_swift();
    char *msg = (char*) malloc(1024);
    sprintf(msg, "\"%s\"\nswift says it's %d chars long\nwhile I say it's %lu chars long\n", swft_str, swift_strlen(swft_str), strlen(swft_str));
    
    swift_printstr(msg);
    swift_printstr(string_from_swift());
    free(swft_str);
    free(msg);

    return 0;
}